### PR TITLE
Added test for namespace controller

### PR DESCRIPTION
**Pull request by anujagrawal699**
**Description:**
This PR introduces tests for namespace controller and adds a test file that significantly enchances its code coverage.

**Additions:**
1. Added a test file namespace_sync_controller_test.go in pkg/controllers/namespace directory.

**Test Coverage**:
Test coverage has been increased from 0% to 41.7%, reflecting the complexity of the controller. To check test coverage, run `go test -coverprofile=coverage.out` in the pkg/controllers/namespace directory.

**What type of PR is this**?
/kind failing-test
/kind feature

**What this PR does / why we need it**:
This PR enhances the testing coverage of the namespace controller, which is crucial for managing namespace operations and ensuring the controller’s stability and functionality. By adding these tests, we ensure that potential edge cases and critical functionalities are properly validated, leading to improved stability and performance of the namespace controller.

**Which issue(s) this PR fixes**:
Fixes a part of #5235 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by anujagrawal699** [comment]
@XiShanYongYe-Chang TestController_SetupWithManager failed during the CI workflow but its running fine locally.
```
 go test -run TestController_SetupWithManager                          
PASS
ok      github.com/karmada-io/karmada/pkg/controllers/namespace 0.094s
```

**Comment by XiShanYongYe-Chang** [comment]
/assign
Hi @anujagrawal699 the ci is failed.

**Comment by anujagrawal699** [comment]
@XiShanYongYe-Chang CI passed.

File: pkg/controllers/namespace/namespace_sync_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+				NamespacedName: types.NamespacedName{
+					Name: tt.namespace.Name,
+				},
+			}
+
+			result, err := c.Reconcile(context.Background(), req)
+			assert.Equal(t, tt.expectedResult, result)
+			assert.Equal(t, tt.expectedError, err != nil)
+
+			if tt.name == "Namespace should be synced" {
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller_test.go:166]: This judgment method is not good. The case name may change. Can you use another method?

File: pkg/controllers/namespace/namespace_sync_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+	_ = workv1alpha1.Install(scheme)
+
+	tests := []struct {
+		name           string
+		namespace      *corev1.Namespace
+		clusters       []clusterv1alpha1.Cluster
+		expectedResult controllerruntime.Result
+		expectedError  bool
+	}{
+		{
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller_test.go:103]: How about adding a few more cases, such as the case where the namespace should not be synchronized, the case where the namepsace cannot be found, and the case where the namespace is in the process of being deleted?

File: pkg/controllers/namespace/namespace_sync_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+			Namespace: names.GenerateExecutionSpaceName(cluster.Name),
+			Name:      names.GenerateWorkName(namespace.Kind, namespace.Name, namespace.Namespace),
+		}, work)
+		assert.NoError(t, err)
+		assert.NotNil(t, work)
+		assert.Equal(t, namespace.Name, work.OwnerReferences[0].Name)
+	}
+}
+
+func TestController_buildWorksWithOverridePolicy(t *testing.T) {
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller_test.go:229]: Is it convenient to combine this test method with TestController_buildWorks into one test method?

File: pkg/controllers/namespace/namespace_sync_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+	assert.NotNil(t, work)
+
+	t.Logf("Work found: %+v", work)
+	t.Logf("Work annotations: %v", work.Annotations)
+	t.Logf("Work spec: %+v", work.Spec)
+
+	assert.NotNil(t, work)
+}
+
+func TestController_SetupWithManager(t *testing.T) {
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller_test.go:310]: Can the following three functions, including this one, be organized into table-driven tests? That would be easier to read and maintain. I do have tests in the current style of testing, but they don't seem to work very well at the moment.

File: pkg/controllers/namespace/namespace_sync_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+		Client:          mgr.GetClient(),
+		EventRecorder:   mgr.GetEventRecorderFor("test-controller"),
+		OverrideManager: overridemanager.New(mgr.GetClient(), mgr.GetEventRecorderFor("test-controller")),
+	}
+
+	err = c.SetupWithManager(mgr)
+	assert.Error(t, err, "SetupWithManager should return an error when schemes are not properly set up")
+}
+
+func TestClusterNamespaceFn(t *testing.T) {
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller_test.go:395]: Sorry, I didn't understand which function body the function tests.

File: pkg/controllers/namespace/namespace_sync_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+			return requests
+		})
+
+	requests := clusterNamespaceFn(context.Background(), nil)
+	assert.Len(t, requests, 2)
+	assert.Contains(t, requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: "namespace1"}})
+	assert.Contains(t, requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: "namespace2"}})
+}
+
+func TestClusterOverridePolicyNamespaceFn(t *testing.T) {
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/namespace/namespace_sync_controller_test.go:431]: ditto.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks a lot @anujagrawal699

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 3 above)
Comment (anujagrawal699) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:229]: I think separate tests will ensure that override policy behavior doesn't interfere with basic functionality. I could create helper functions for common setup code if you say.

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 5 above)
Comment (anujagrawal699) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:395]: The two you mentioned are clusterNamespaceFn and clusterOverridePolicyNamespaceFn which are MapFuncs used in the controller's SetupWithManager() method to handle events related to Cluster and ClusterOverridePolicy resources.

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 4 above)
Comment (anujagrawal699) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:310]: Yeah, that would be more readable and maintainable.

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 1 above)
Comment (anujagrawal699) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:166]: Okay, onto it.

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 2 above)
Comment (anujagrawal699) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:103]: Yes, it would also improve test coverage.

**Comment by anujagrawal699** [comment]
@XiShanYongYe-Chang PTAL

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 5 above)
Comment (XiShanYongYe-Chang) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:395]: I understand what you are saying, but where is the code guarded by these two test methods?

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 3 above)
Comment (XiShanYongYe-Chang) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:229]: You can modify it to see the effect.

File: pkg/controllers/namespace/namespace_sync_controller_test.go (hunk 5 above)
Comment (anujagrawal699) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:395]: Okay, i got it. I think i need to refractor the whole  controller struct. Thank You for pointing out.

Comment (anujagrawal699) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:395]: @XiShanYongYe-Chang I'm not able to test them in isolation. I think we should ahead without them.

Comment (XiShanYongYe-Chang) [reply pkg/controllers/namespace/namespace_sync_controller_test.go:395]: Yes, I don't think they need to be tested.

**Comment by NishantBansal2003** [comment]
Hey @anujagrawal699, could you please review [my PR](https://github.com/karmada-io/karmada/pull/5390) as well? Thanks!

**Comment by anujagrawal699** [comment]
> Hey @anujagrawal699, could you please review [my PR](https://github.com/karmada-io/karmada/pull/5390) as well? Thanks!

Sure!

**Comment by liangyuanpeng** [comment]
/ok-to-test 
then you can run `/retest` to rerun github workflow when all workflow is completed

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks
/lgtm
/approve

**Comment by anujagrawal699** [comment]
> Thanks
> /lgtm
> /approve

Thanks for approval. I could have squashed the commits.

**Comment by XiShanYongYe-Chang** [comment]
> I could have squashed the commits.

I forgot again.

**Comment by anujagrawal699** [comment]
> > I could have squashed the commits.
> 
> I forgot again.

Its Okay. Thanks!
